codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/EnumGenerator.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.go.codegen;
+
+import java.util.Locale;
+import java.util.Map;
+import software.amazon.smithy.codegen.core.Symbol;
+import software.amazon.smithy.codegen.core.SymbolProvider;
+import software.amazon.smithy.model.shapes.StringShape;
+import software.amazon.smithy.model.traits.EnumConstantBody;
+import software.amazon.smithy.model.traits.EnumTrait;
+import software.amazon.smithy.utils.StringUtils;
+
+/**
+ * Renders enums and their constants.
+ */
+final class EnumGenerator implements Runnable {
+
+    private final SymbolProvider symbolProvider;
+    private final GoWriter writer;
+    private final StringShape shape;
+
+    EnumGenerator(SymbolProvider symbolProvider, GoWriter writer, StringShape shape) {
+        this.symbolProvider = symbolProvider;
+        this.writer = writer;
+        this.shape = shape;
+    }
+
+    @Override
+    public void run() {
+        Symbol symbol = symbolProvider.toSymbol(shape);
+        EnumTrait enumTrait = shape.expectTrait(EnumTrait.class);
+
+        writer.write("type $L string", symbol.getName()).write("");
+
+        writer.write("// Enum values for $L", symbol.getName()).openBlock("const (", ")", () -> {
+            for (Map.Entry<String, EnumConstantBody> entry : enumTrait.getValues().entrySet()) {
+                StringBuilder labelBuilder = new StringBuilder(symbol.getName());
+                String name = entry.getValue().getName().orElse(entry.getKey());
```

Comment:
Do valid enum values (aside from the whitespace which is cleaned up below) work 100% as parts of Go's identifiers?

Reply:
If the enum value has any non Unicode letter, or number character those have to be stripped as well. And enum must start with an Unicode letter character. So things like `__` prefix must be stripped out or the type will not be public.

Though since Smithy allows naming of enum values e.g. map. this should resolve the issue. As long as Smithy's name keys are not too permissive.

Reply:
Not all enums will have names, even if we recommend it. We need to make sure all cases for valid values for these are handled.

Given that the name components are appended to the symbol name, the `__` prefix issue shouldn't be a problem specifically.

Reply:
The prefix issue would be the only real problem. There's standard bits like Json -> JSON but that's not critical.